Add fast sums and sums of squares over quantized ranges to QuantizedOpKernels.cpp

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35329 add quantized layer norm implementation
* **#35693 Add fast sums and sums of squares over quantized ranges to QuantizedOpKernels.cpp**

Summary:

Adds utility functions to quantized int types to calculate
horizontal sums and sums of squares using avx2 intrinsics.

This is useful for quantized implementations of various normalization
layers (LayerNorm, GroupNorm, InstanceNorm), where we need to calculate
the mean and variance of a layer of quantized ints.

Test Plan:

Adhoc c++ tester for the correctness of the avx2 functions:
https://gist.github.com/vkuzo/0380f450793cd5c05abbeacb6d3883ae
Run with:
```
-lstdc++ -mavx2 -lm -ldl -o main main.cpp && ./main
```

The integration bits and performance will be tested in the next PR in the stack
where we will hook quantized Layernorm to use this.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20768804](https://our.internmc.facebook.com/intern/diff/D20768804)

vkuzo committed:
Add avx2 integer horizontal sum and sum of squares to vec256 qint types

Summary:

Adds utility functions to quantized int types of vec256 to calculate
horizontal sums and sums of squares using avx2 intrinsics.

This is useful for quantized implementations of various normalization
layers (LayerNorm, GroupNorm, InstanceNorm), where we need to calculate
the mean and variance of a layer of quantized ints.

Test Plan:

Adhoc c++ tester for the correctness of the avx2 functions:
https://gist.github.com/vkuzo/0380f450793cd5c05abbeacb6d3883ae
Run with:
```
-lstdc++ -mavx2 -lm -ldl -o main main.cpp && ./main
```

The integration bits and performance will be tested in the next PR in the stack
where we will hook quantized Layernorm to use this.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/cpu/vec256/vec256_qint.h:
```diff
@@ -154,6 +154,223 @@ inline void __attribute__((always_inline)) QuantizeAvx2(
+
+// horizontal sums signed i64, overflow unsafe
+// x = (y3, y2, y1, y0)
+int64_t custom_mm256_hsum_epi64_ignore_overflow(__m256i x) {
```
p.s. I tried to put these in a separate header but got a bunch of linking errors with "multiple functions of the same name" defined across various build flags.  Let me know if there is something special that needs to be done for a new header in this dir.

jamesr66a commented on aten/src/ATen/cpu/vec256/vec256_qint.h:
```diff
@@ -154,6 +154,223 @@ inline void __attribute__((always_inline)) QuantizeAvx2(
+
+// horizontal sums signed i8
+// x = (x31, ..., x0)
+int custom_mm256_hsum_epi8(__m256i x) {
```
The return value is immediately getting widened to int64_t at the callsite. Should we make that explicit in the return type here (and below?)

jamesr66a reviewed (commented):
Seems reasonable to me, but you might want to name the hsum functions something that explicitly mentions the widening behavior to distinguish those functions from something like _mm256_hadd_epi{16}, which would overflow

cc @dskhudia can you take a look as well?

vkuzo committed:
Update on "Add avx2 integer horizontal sum and sum of squares to vec256 qint types"


Summary:

Adds utility functions to quantized int types of vec256 to calculate
horizontal sums and sums of squares using avx2 intrinsics.

This is useful for quantized implementations of various normalization
layers (LayerNorm, GroupNorm, InstanceNorm), where we need to calculate
the mean and variance of a layer of quantized ints.

Test Plan:

Adhoc c++ tester for the correctness of the avx2 functions:
https://gist.github.com/vkuzo/0380f450793cd5c05abbeacb6d3883ae
Run with:
```
-lstdc++ -mavx2 -lm -ldl -o main main.cpp && ./main
```

The integration bits and performance will be tested in the next PR in the stack
where we will hook quantized Layernorm to use this.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dskhudia commented on aten/src/ATen/cpu/vec256/vec256_qint.h:
```diff
@@ -154,6 +154,223 @@ inline void __attribute__((always_inline)) QuantizeAvx2(
+  // (x15, ..., x0), int8
+  const __m128i xHalf2 = _mm256_castsi256_si128(x);
+  // (x31, ... x16), int16
+  const __m256i xHalf1_64 = _mm256_cvtepu8_epi16(xHalf1);
+  // (x15, ..., x0), int16
+  const __m256i xHalf2_64 = _mm256_cvtepu8_epi16(xHalf2);
```
Any reason not to use _mm256_hadd_epi16 on xHalf1_64 and xHalf2_64 and then forming a tree of hadds?

vkuzo replied:
just to clarify, you mean why not call custom_mm256_hsum_epu16_overflow from here?  If yes - it would be slower (can't remember by how much but I did measure it in my adhoc tester and it was significant) because that functions widens the inputs again, and we only need to widen once to ensure no overflow.

vkuzo replied:
I think it was more than 20% slower, but can run again if needed for the exact #

dskhudia replied:
I meant the horizontal add intrinsic itself on 16-bit values.

vkuzo replied:
ah, I didn't know about `hadd` (was searching for `hsum`).  Thanks for the tip! Yeah, that should improve things, along with your other suggestion - will check it out and benchmark

vkuzo committed:
Update on "Add avx2 integer horizontal sum and sum of squares to vec256 qint types"


Summary:

Adds utility functions to quantized int types of vec256 to calculate
horizontal sums and sums of squares using avx2 intrinsics.

This is useful for quantized implementations of various normalization
layers (LayerNorm, GroupNorm, InstanceNorm), where we need to calculate
the mean and variance of a layer of quantized ints.

Test Plan:

Adhoc c++ tester for the correctness of the avx2 functions:
https://gist.github.com/vkuzo/0380f450793cd5c05abbeacb6d3883ae
Run with:
```
-lstdc++ -mavx2 -lm -ldl -o main main.cpp && ./main
```

The integration bits and performance will be tested in the next PR in the stack
where we will hook quantized Layernorm to use this.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20768804](https://our.internmc.facebook.com/intern/diff/D20768804)

[ghstack-poisoned]

dskhudia commented:
hsum for int8 and uint8 can be combined using a template since the code is mostly the same. Similarly hsum_sq for int8 and uint8. Other than this, it looks good to me.

dskhudia commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -152,6 +152,226 @@ Tensor qcat_nhwc_kernel(
+    );
+  }
+
+  alignas(64) int32_t temp[8];
+  _mm256_store_si256(reinterpret_cast<__m256i*>(temp), sum_v);
+  for (int k = 0; k < 8; ++k) {
+    row_sum += temp[k];
+  }
```
If you are feeling adventurous, you can do this part using _mm256_hadd_epi32 and the remainder part below using mask instructions. For example, see use of masking in remainder loop in https://github.com/pytorch/FBGEMM/blob/master/src/QuantUtilsAvx2.cc#L91-L97

vkuzo commented:
> hsum for int8 and uint8 can be combined using a template since the code is mostly the same. Similarly hsum_sq for int8 and uint8. Other than this, it looks good to me.

hmm, not sure if this is worth it, as unless we templatize all three types we'll have to branch at the callsites

vkuzo committed:
Update on "Add fast sums and sums of squares over quantized ranges to QuantizedOpKernels.cpp"


Summary:

Adds utility functions to quantized int types to calculate
horizontal sums and sums of squares using avx2 intrinsics.

This is useful for quantized implementations of various normalization
layers (LayerNorm, GroupNorm, InstanceNorm), where we need to calculate
the mean and variance of a layer of quantized ints.

Test Plan:

Adhoc c++ tester for the correctness of the avx2 functions:
https://gist.github.com/vkuzo/0380f450793cd5c05abbeacb6d3883ae
Run with:
```
-lstdc++ -mavx2 -lm -ldl -o main main.cpp && ./main
```

The integration bits and performance will be tested in the next PR in the stack
where we will hook quantized Layernorm to use this.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20768804](https://our.internmc.facebook.com/intern/diff/D20768804)

[ghstack-poisoned]

vkuzo committed:
Update on "Add fast sums and sums of squares over quantized ranges to QuantizedOpKernels.cpp"


Summary:

Adds utility functions to quantized int types to calculate
horizontal sums and sums of squares using avx2 intrinsics.

This is useful for quantized implementations of various normalization
layers (LayerNorm, GroupNorm, InstanceNorm), where we need to calculate
the mean and variance of a layer of quantized ints.

Test Plan:

Adhoc c++ tester for the correctness of the avx2 functions:
https://gist.github.com/vkuzo/0380f450793cd5c05abbeacb6d3883ae
Run with:
```
-lstdc++ -mavx2 -lm -ldl -o main main.cpp && ./main
```

The integration bits and performance will be tested in the next PR in the stack
where we will hook quantized Layernorm to use this.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20768804](https://our.internmc.facebook.com/intern/diff/D20768804)

[ghstack-poisoned]
